common/src/main/java/io/netty/util/concurrent/DefaultPromise.java

```diff
@@ -790,96 +693,30 @@ private static void notifyProgressiveListener0(
+    private static boolean isCancelled0(Object result) {
+        return result instanceof CauseHolder && ((CauseHolder) result).cause instanceof CancellationException;
     }
 
-    protected StringBuilder toStringBuilder() {
-        StringBuilder buf = new StringBuilder(64)
-            .append(StringUtil.simpleClassName(this))
-            .append('@')
-            .append(Integer.toHexString(hashCode()));
-
-        Object result = this.result;
-        if (result == SUCCESS) {
-            buf.append("(success)");
-        } else if (result == UNCANCELLABLE) {
-            buf.append("(uncancellable)");
-        } else if (result instanceof CauseHolder) {
-            buf.append("(failure: ")
-               .append(((CauseHolder) result).cause)
-               .append(')');
-        } else if (result != null) {
-            buf.append("(success: ")
-               .append(result)
-               .append(')');
-        } else {
-            buf.append("(incomplete)");
-        }
-
-        return buf;
+    private static boolean isDone0(Object result) {
+        return result != null && result != UNCANCELLABLE;
     }
 
-    private final class LateListeners extends ArrayDeque<GenericFutureListener<?>> implements Runnable {
-
-        private static final long serialVersionUID = -687137418080392244L;
-
-        LateListeners() {
-            super(2);
-        }
-
-        @Override
-        public void run() {
-            final EventExecutor executor = executor();
-            if (listeners == null || executor == ImmediateEventExecutor.INSTANCE) {
-                for (;;) {
-                    GenericFutureListener<?> l = poll();
-                    if (l == null) {
-                        break;
-                    }
-                    notifyListener0(DefaultPromise.this, l);
-                }
-            } else {
-                // Reschedule until the initial notification is done to avoid the race condition
-                // where the notification is made in an incorrect order.
-                safeExecute(executor, this);
-            }
+    private static final class CauseHolder {
+        final Throwable cause;
+        CauseHolder(Throwable cause) {
+            this.cause = cause;
         }
     }
 
-    private final class LateListenerNotifier implements Runnable {
-        private GenericFutureListener<?> l;
-
-        LateListenerNotifier(GenericFutureListener<?> l) {
-            this.l = l;
-        }
-
-        @Override
-        public void run() {
-            LateListeners lateListeners = DefaultPromise.this.lateListeners;
-            if (l != null) {
-                if (lateListeners == null) {
-                    DefaultPromise.this.lateListeners = lateListeners = new LateListeners();
-                }
-                lateListeners.add(l);
-                l = null;
-            }
-
-            lateListeners.run();
+    private static void safeExecute(EventExecutor executor, Runnable task) {
```

Comment:
@Scottmitch maybe not related to this PR but we kind of have the same method in other place in Netty as well. Maybe factor out in a utility method ?